File: frontend/web/services/connections/connect.ts
import { isAxiosError } from "axios";
import { StatusCodes } from "http-status-codes";
import { axiosInstance } from "@/services";
import { SERVICE_ERROR_UNKNOWN } from "@/config/services";
import { ServiceReturn } from "@/types/api";

type CheckConnectionReturn = {
	oauthUrl: string | null;
};
const connect = async (serviceId: string, scopes: string[]): Promise<ServiceReturn<CheckConnectionReturn>> => {
	if (scopes.length === 0) {
		// Account doesn't need connection
		return { data: { oauthUrl: "" }, error: undefined };
	}
	try {
		const response = await axiosInstance.post<CheckConnectionReturn>(`/connections/${serviceId}/connect`, {
			scopes,
		});

		return { data: response.data, error: undefined };
	} catch (error) {
		if (isAxiosError(error)) {
			if (error.response?.status === StatusCodes.CONFLICT) {
				// Account already exists with the necessary permissions
				return { data: { oauthUrl: null }, error: undefined };
			}
			if (error.response?.status === StatusCodes.FORBIDDEN) {
				// Account doesn't need connection
				return { data: { oauthUrl: "" }, error: undefined };
			}
		}
		return { data: null, error: SERVICE_ERROR_UNKNOWN };
	}
};

// eslint-disable-next-line import/prefer-default-export
export { connect };

File: frontend/mobile/lib/services/connections/connect.dart
import 'package:area_mobile/services/dio.dart';
import 'package:area_mobile/types/services.dart';
import 'package:dio/dio.dart';

Future<ServiceReturn<String>> connect(
    String serviceId, List<String> scopes) async {
  if (scopes.isEmpty) {
    // Account doesn't need connection
    return const ServiceReturn(data: "");
  }
  try {
    final response = await dio.post('/connections/$serviceId/connect',
        data: {"scopes": scopes.toList()});
    return ServiceReturn(data: response.data["oauthUrl"]);
  } catch (e) {
    final error = e as DioException;
    if (error.response?.statusCode == 409) {
      // Account already exists with the necessary permissions
      return const ServiceReturn(data: null);
    }
    if (error.response?.statusCode == 403) {
      // Account doesn't need connection
      return const ServiceReturn(data: "");
    }
    return ServiceReturn(error: e.toString());
  }
}
